### PR TITLE
fix(core): Add stage prefix to stack name shortening process

### PR DESCRIPTION
**Pull request by QustodioPablo**
There is an issue in the stack name generation process where the prefix generated from assembly's stage name is not taken into account when shortening a stack name to meet the requirement of being equal or less than 128 characters longs.

This can lead to generating an invalid stack name greater than 128 characters: since the stack name is shortened to 128 characters, when the prefix is added, the limit is exceeded.

Current solution:
- Adding a feature flag
  - With the feature flag on, the prefix is processed within the generateUniqueName function.
  - With the feature flag off, stack name generation is not changed

Fixes https://github.com/aws/aws-cdk/issues/23628

NOTE: This PR was previously opened, but it was merged before I was able to add a feature flag, which ended up adding breaking changes and the changes of the PR were rolled back. Old PR: https://github.com/aws/aws-cdk/pull/24443

### CODE REVIEW

**Comment by QustodioPablo** [comment]
Exemption Request: The previous PR was exempt from adding integration tests

**Comment by QustodioPablo** [comment]
Clarification Request: Can a maintainer confirm if this PR needs integration tests?

**Review by crbernal — APPROVED** [review]

**Review by andypapas — APPROVED** [review]

**Review by ygneo — APPROVED** [review]

**Comment by ygneo** [comment]
👍

**Review by ericgimenezgalera — APPROVED** [review]

**Review by MariposaGentil — APPROVED** [review]

**Review by cas-jc — APPROVED** [review]

**Review by lauralongares — APPROVED** [review]

**Review by lauralongares — APPROVED** [review]

File: packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts [hunk 1]
```diff
@@ -59,7 +67,7 @@ export function makeUniqueResourceName(components: string[], options: MakeUnique
   // in order to support transparent migration of cloudformation templates to the CDK without the
   // need to rename all resources.
   if (components.length === 1) {
-    const topLevelResource = removeNonAllowedSpecialCharacters(components[0], separator, options.allowedSpecialCharacters);
+    const topLevelResource = prefix + removeNonAllowedSpecialCharacters(components[0], separator, options.allowedSpecialCharacters);
```
Comment (corymhall) [inline packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts:70]: Shouldn't this prefix also be included in the `removeNonAllowedSpecialCharacters`?
```suggestion
    const topLevelResource = removeNonAllowedSpecialCharacters(prefix + components[0], separator, options.allowedSpecialCharacters);
```

File: packages/aws-cdk-lib/core/lib/stack.ts [hunk 2]
```diff
@@ -1644,9 +1649,14 @@ export function rootPathTo(construct: IConstruct, ancestor?: IConstruct): IConst
  * has only one component. Otherwise we fall back to the regular "makeUniqueId"
  * behavior.
  */
-function makeStackName(components: string[]) {
-  if (components.length === 1) { return components[0]; }
-  return makeUniqueResourceName(components, { maxLength: 128 });
+function makeStackName(components: string[], prefix: string='') {
+  if (components.length === 1) {
+    const stack_name = prefix + components[0];
```
Comment (corymhall) [inline packages/aws-cdk-lib/core/lib/stack.ts:1654]: why not just let `makeUniqueResourceName` do this?

File: packages/aws-cdk-lib/core/test/stage.test.ts
Comment (corymhall) [inline packages/aws-cdk-lib/core/test/stage.test.ts:1]: Can you add a couple more tests:

- stage & stack with 1 component and < 128
- stage and stack with > 1 component and < 128
- stage and stack with > 1 component and > 128

File: packages/aws-cdk-lib/cx-api/lib/features.ts [hunk 3]
```diff
@@ -789,6 +790,21 @@ export const FLAGS: Record<string, FlagInfo> = {
...
+
+  //////////////////////////////////////////////////////////////////////
+  [INCLUDE_PREFIX_IN_UNIQUE_NAME_GENERATION]: {
+    type: FlagType.BugFix,
+    summary: 'Include the stack prefix in the stack name generation process',
+    detailsMd: `
+      This flag prevents the prefix of a stack from making the stack's name longer than the 128 character limit.
+
+      If the flag is set, the prefix is included in the stack name generation process.
+      If the flag is not set, then the prefix of the stack is prepended to the generated stack name.
```
Comment (corymhall) [inline packages/aws-cdk-lib/cx-api/lib/features.ts:802]: Can you add a note here explaining what will happen if you enable this feature flag on an existing stack.

**Review by corymhall — CHANGES_REQUESTED** [review]
@QustodioPablo just a couple of questions, otherwise looks good.

File: packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts (hunk 1 above)
Comment (QustodioPablo) [reply packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts:70]: This is done to keep the new name as close as possible to the original behavior. When a shortened stack name had a prefix, the final result would be "Prefix-ShortenedStackName". If the prefix is included in the removal of special characters, the name generated would be "PrefixShortenedStackName".

File: packages/aws-cdk-lib/core/lib/stack.ts (hunk 2 above)
Comment (QustodioPablo) [reply packages/aws-cdk-lib/core/lib/stack.ts:1654]: This is done also to make the stack names to be as close as possible to the original behavior. If we go through the makeUniqueResourceName, we would lose all non allowed special character (dashes for example) from the stack name if the component has any in its name. So we don't want single component stacks to go through the makeUniqueResourceName in order to preserve its original name.

File: packages/aws-cdk-lib/core/test/stage.test.ts
Comment (QustodioPablo) [reply packages/aws-cdk-lib/core/test/stage.test.ts:1]: All of them with the feature flag on, right?

File: packages/aws-cdk-lib/cx-api/lib/features.ts (hunk 3 above)
Comment (QustodioPablo) [reply packages/aws-cdk-lib/cx-api/lib/features.ts:802]: Sure!

File: packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts (hunk 1 above)
Comment (corymhall) [reply packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts:70]: ah yeah you are right 👍

File: packages/aws-cdk-lib/core/test/stage.test.ts
Comment (corymhall) [reply packages/aws-cdk-lib/core/test/stage.test.ts:1]: yep!

Comment (QustodioPablo) [reply packages/aws-cdk-lib/core/test/stage.test.ts:1]: One of the suggested tests found a case where the name generated would be different when using this FF, the same scenario that caused the previous PR to be reverted: The dash from the prefix was being removed when a multi component stack is processed in the makeUniqueResourceName function.

Fixed it

File: packages/aws-cdk-lib/cx-api/lib/features.ts (hunk 3 above)
Comment (QustodioPablo) [reply packages/aws-cdk-lib/cx-api/lib/features.ts:801]: Done

**Comment by QustodioPablo** [comment]
@TheRealAmazonKendra I reopened this PR adding a feature flag, could you add an exception for integration tests as in the previous PR?

**Review by otaviomacedo — DISMISSED** [review]

**Pull request merged**